platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/Bucket.java

```diff
@@ -246,242 +217,262 @@ public long findValue(final int keyHashCode, final K key, final long notFoundVal
-     * "entryOffset" is the entry offset in bytes in the bucket buffer, "entryIndex" is the
-     * entry index in the array of entries, and "entryValue" is the entry value. If no entity
-     * is found, "found" is false, "entryOffset" is the total size of the bucket buffer,
-     * "entryIndex" and "entryValue" are undefined.
+     * A single entry in a bucket, which contains key hash code, value (usually, path), and
+     * full serialized key. A bucket may contain multiple such entries.
+     *
+     * <p>This class would be a record, if it was immutable. However, when a value is updated
+     * in a bucket, and a bucket entry already exists for the same key, instead of creating
+     * a new entry, we just update the value in the existing entry.
      */
-    private record FindResult(int entryOffset, int entryIndex, boolean found, long entryValue) {}
+    private class BucketEntry {
+
+        /** Key hash code */
+        private int hashCode;
+        /** Long value */
+        private long value;
+        /** Key */
+        private K key;
+        /** Serialized key */
+        private BufferedData keyBytes;
+
+        /** Creates new bucket entry from hash code, value, and serialized key bytes */
+        public BucketEntry(final int hashCode, final long value, @NonNull final K key) {
+            this.hashCode = hashCode;
+            this.value = value;
+            this.key = key;
+            this.keyBytes = null;
+        }
+
+        /** Creates new bucket entry by reading its fields from the given protobuf buffer */
+        public BucketEntry(final ReadableSequentialData entryData) {
+            // defaults
+            hashCode = 0;
+            value = 0;
+            key = null;
+            keyBytes = null;
+
+            // read fields
+            while (entryData.hasRemaining()) {
+                final int tag = entryData.readVarInt(false);
+                final int fieldNum = tag >> TAG_FIELD_OFFSET;
+                if (fieldNum == FIELD_BUCKETENTRY_HASHCODE.number()) {
+                    hashCode = entryData.readInt();
+                } else if (fieldNum == FIELD_BUCKETENTRY_VALUE.number()) {
+                    value = entryData.readLong();
+                } else if (fieldNum == FIELD_BUCKETENTRY_KEYBYTES.number()) {
+                    final int bytesSize = entryData.readVarInt(false);
+                    long oldLimit = entryData.limit();
+                    entryData.limit(entryData.position() + bytesSize);
+                    key = keySerializer.deserialize(entryData);
+                    entryData.limit(oldLimit);
+                } else {
+                    throw new IllegalArgumentException("Unknown bucket entry field: " + fieldNum);
+                }
+            }
 
-    /** Get bucket buffer for tests */
-    ByteBuffer getBucketBuffer() {
-        return bucketBuffer;
+            // check required fields
+            if ((key == null) && (keyBytes == null)) {
```

Comment:
Why check keyBytes?  It's always null here (set on 375, and never changed in this constructor)...